### PR TITLE
[DNM] parser: reduce memory allocation

### PR DESCRIPTION
**Pull request by tiancaiamao**
change the yacc code generator to make less make([]xxx) call.
but yyParse still cause too much allocations after this change.

tidb-bench  before vs after:
```
insert test data finished ... elapse : 34.563343s tps:28932.386438
insert test data finished ... elapse : 28.859153s tps:34651.051825
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -54,7 +54,7 @@ install:
 TEMP_FILE = temp_parser_file
 
 parser:
-	go get github.com/qiuyesuifeng/goyacc
+	go get github.com/tiancaiamao/goyacc
```
Comment (shenli) [inline Makefile]: Move the repo into pingcap account.

File: evaluator/evaluator_test.go [hunk 2]
```diff
@@ -42,7 +42,7 @@ type testEvaluatorSuite struct {
 }
 
 func parseExpr(c *C, expr string) ast.ExprNode {
-	s, err := parser.ParseOneStmt("select "+expr, "", "")
+	s, err := parser.New().ParseOneStmt("select "+expr, "", "")
```
Comment (shenli) [inline evaluator/evaluator_test.go]: It is better to create a global parser variable to reduce test time.

File: executor/executor_simple.go [hunk 3]
```diff
@@ -355,7 +354,6 @@ func (e *SimpleExec) createStatisticsForTable(tn *ast.TableName) error {
 // collectSamples collects sample from the result set, using Reservoir Sampling algorithm.
 // See https://en.wikipedia.org/wiki/Reservoir_sampling
 func (e *SimpleExec) collectSamples(result ast.RecordSet) (count int64, samples []*ast.Row, err error) {
-	ran := rand.New(rand.NewSource(time.Now().UnixNano()))
```
Comment (shenli) [inline executor/executor_simple.go]: This will not be a problem. For each table, it will be called only once.

File: executor/prepared.go [hunk 4]
```diff
@@ -121,7 +121,7 @@ func (e *PrepareExec) DoPrepare() {
 		}
 	}
 	charset, collation := variable.GetCharsetInfo(e.Ctx)
-	stmts, err := parser.Parse(e.SQLText, charset, collation)
+	stmts, err := parser.New().Parse(e.SQLText, charset, collation)
```
Comment (shenli) [inline executor/prepared.go]: Get parser from e.Ctx.

File: parser/yy_parser.go [hunk 5]
```diff
@@ -58,9 +58,21 @@ func handleMySQLSpecificCode(sql string) string {
 	return specCodePattern.ReplaceAllStringFunc(sql, trimComment)
 }
 
+// Parser provide the parser instance
+type Parser struct {
+	cache []yySymType
+}
+
+// New return a Parser object
```
Comment (shenli) [inline parser/yy_parser.go]: returns

File: parser/yy_parser.go [hunk 6]
```diff
@@ -58,9 +58,21 @@ func handleMySQLSpecificCode(sql string) string {
 	return specCodePattern.ReplaceAllStringFunc(sql, trimComment)
 }
 
+// Parser provide the parser instance
```
Comment (shenli) [inline parser/yy_parser.go]: The comment is not clear.

Comment (zimulala) [inline parser/yy_parser.go]: provide -> provides, and add .

File: parser/yy_parser.go (hunk 5 above)
Comment (zimulala) [inline parser/yy_parser.go]: return -> returns, and add .

**Comment by shenli** [comment]
The performance result is very promising. Can you run sysbench on the new branch?

File: executor/prepared.go [hunk 7]
```diff
@@ -121,7 +122,15 @@ func (e *PrepareExec) DoPrepare() {
...
 	}
 	charset, collation := variable.GetCharsetInfo(e.Ctx)
-	stmts, err := parser.Parse(e.SQLText, charset, collation)
+	var (
+		stmts []ast.StmtNode
+		err   error
+	)
+	if sqlParser, ok := e.Ctx.(sqlexec.SQLParser); ok {
+		stmts, err = sqlParser.ParseSQL(e.SQLText, charset, collation)
+	} else {
```
Comment (shenli) [inline executor/prepared.go]: Why we need the else?

Comment (tiancaiamao) [inline executor/prepared.go]: if one day context change to other implement, or been wrapped in another struct, we won't meet a strange panic.
It's not good to make strong assumption of a particular implementation.

File: session.go [hunk 8]
```diff
@@ -272,7 +274,8 @@ func (s *session) Retry() error {
 // ExecRestrictedSQL implements SQLHelper interface.
 // This is used for executing some restricted sql statements.
 func (s *session) ExecRestrictedSQL(ctx context.Context, sql string) (ast.RecordSet, error) {
-	rawStmts, err := Parse(ctx, sql)
+	charset, collation := getCtxCharsetInfo(s)
+	rawStmts, err := s.parser.Parse(sql, charset, collation)
```
Comment (shenli) [inline session.go]: You can use ParseSQL.

File: session.go [hunk 9]
```diff
@@ -380,11 +383,18 @@ func (s *session) ShouldAutocommit(ctx context.Context) bool {
...
 }
 
+func (s *session) ParseSQL(sql, charset, collation string) ([]ast.StmtNode, error) {
+	return s.parser.Parse(sql, charset, collation)
+}
+
 func (s *session) Execute(sql string) ([]ast.RecordSet, error) {
-	rawStmts, err := Parse(s, sql)
+	charset, collation := getCtxCharsetInfo(s)
+	rawStmts, err := s.parser.Parse(sql, charset, collation)
```
Comment (shenli) [inline session.go]: s.ParseSQL

File: util/sqlexec/restricted_sql_executor.go [hunk 10]
```diff
@@ -34,3 +34,10 @@ type RestrictedSQLExecutor interface {
 	// ExecRestrictedSQL run sql statement in ctx with some restriction.
 	ExecRestrictedSQL(ctx context.Context, sql string) (ast.RecordSet, error)
 }
+
+// SQLParser is a bit tricky. We could run parser.New() to get a parser object, and then run Parse method on it.
+// But a session already has a parser bind in it, so we define this interface to avoid allocate a new parser.
+// See session.SQLParser for more information.
+type SQLParser interface {
```
Comment (shenli) [inline util/sqlexec/restricted_sql_executor.go]: We should first describe the function of this interface. But the comment only contains the detail of its implementation.

**Comment by shenli** [comment]
LGTM

**Comment by shenli** [comment]
@coocood PTAL

**Comment by coocood** [comment]
LGTM

**Pull request closed without merging**